hyperapp.d.ts

```diff
@@ -0,0 +1,128 @@
+export as namespace Hyperapp
+
+/** @namespace [Virtual DOM] */
+
+/** The Virtual DOM representation of an Element
+ *
+ * @memberOf [Virtual DOM]
+*/
+export interface VirtualNode<Data> {
+  tag: string
+  data: Data
+  children: VirtualNodeChild<{} | null>[]
+}
+
+/** In the Virtual DOM a Child could be either a Virtual Node or a string
+ *
+ * @memberOf [Virtual DOM]
+*/
+export type VirtualNodeChild<Data> = VirtualNode<Data> | string
+
+export interface Component<Data> {
+  /** A Component is a function that return a custom Virtual Node
+   *
+   * @memberOf [Virtual DOM]
+  */
+  (data: Data, children: VirtualNodeChild<{} | null>[]): VirtualNode<Data>
+}
+
+/** The soft way to create a Virtual Node
+ * @param tag       Either a tag name e.g. 'div'. Or a Component function
+ * @param data      Any valid HTML atributes, events, styles, and meta data
+ * @param children  The children of the VirtualNode
+ *
+ * @memberOf [Virtual DOM]
+*/
+export function h<Data>(
+  tag: Component<Data> | string,
+  data?: Data,
+  children?:
+    | VirtualNodeChild<{} | null>[]
+    | VirtualNodeChild<{} | null>
+    | number
+): VirtualNode<Data>
+
+/** @namespace [Application] */
+
+export interface State {}
+
+export interface Thunk {
+  (update: Function): any | null | void
+}
+
+export type ActionsResult<State extends Hyperapp.State> =
+  | Partial<State>
+  | Thunk
+  | {}
+  | null
+  | void
+
+
+export type ActionsCallers<
+  State extends Hyperapp.State,
+  Actions extends Hyperapp.Actions<State>
+> = {
+  [P in keyof Actions]: (
+    data: any
+  ) => ActionsResult<State> | ActionsCallers<State, Actions>
+}
+
+export interface Actions<State extends Hyperapp.State> {
+  [action: string]:
+    | Actions<State>
+    | ((
+        state: State,
+        actions: ActionsCallers<State, Actions<State>>,
+        data: any
+      ) => ActionsResult<State>)
+}
+
+export interface View<
+  State extends Hyperapp.State,
+  Actions extends Hyperapp.Actions<State>
+> {
+  (state: State, actions: ActionsCallers<State, Actions>): VirtualNode<{}>
+}
+
+export interface Events<
+  State extends Hyperapp.State,
+  Actions extends Hyperapp.Actions<State>
+> {}
```

Comment:
@Swizz What about an event third `data` argument?

Reply:
This is not arguments but the generics. Here is no interface for Events yet